cpp/src/connection.cpp

```diff
@@ -192,6 +193,13 @@ std::vector<symbol> connection::desired_capabilities() const {
     return get_multiple<std::vector<symbol> >(caps);
 }
 
+std::map<symbol, value> connection::properties() const {
+    map<symbol, value> props(pn_connection_remote_properties(pn_object()));
+    std::map<symbol, value> props_ret;
+    get(props, props_ret);
+    return props_ret;
```

Comment:
I think you mean:
```
    value props(pn_connection_remote_properties(pn_object()));
    if (props.empty()) {
      return std::map<symbol, value>();
    } else {
      return get<std::map<symbol, value> >(props);
    }

```
This avoids building an entirely unecessary proton::map just to throw it away immediately.

Reply:
Thank you for your comments.  I agree my code is wasteful in the case where there are no properties and I will rectify that.  However, I think it was necessary to create an intermediate `proton::map` in order to get the `decoder` to process all the properties data, rather than just the first key - buy I may be mis-remembering things so  will have another go.

Reply:
It's not the no properties case that is wasteful, it's the with properties case. There it builds the proton::map unnecessarily and throws it away. My replacement code doesn't use a proton::map at all. The whole process is still annoyingly heavyweight, but at least this isn't going to be on the hot path in any sensible use.
My guess is that you used code like mine initially but without the null check (because it's the parallel to the capabilities code). But that doesn't work and fails the unit test you provided. 
As the code passes the unit test I'm reasonably sure it doesn't just decode the first element!

Reply:
Thanks for your elucidation.  I was mis-remembering my earlier attempts but your suggestion helped me track down why my earlier code had been failing the unit tests  (memory management in the non-NULL case).
I have ended up with a slight variation on your suggestion due to a preference for single return points, but I think it is as efficient - and an improvement on my first submission.